p2p/server.py

```diff
@@ -0,0 +1,136 @@
+import asyncio
+import logging
+import secrets
+
+from typing import Tuple
+from eth_keys import datatypes
+
+from evm.db.chain import AsyncChainDB
+from p2p.auth import (
+    decode_auth_plain,
+    decode_auth_eip8,
+    HandshakeResponder,
+)
+from p2p.cancel_token import CancelToken
+from p2p.constants import (
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+)
+from p2p.discovery import DiscoveryProtocol
+from p2p.ecies import ecdh_agree
+from p2p.exceptions import OperationCancelled
+from p2p.kademlia import (
+    Address,
+    Node,
+)
+from p2p.peer import (
+    ETHPeer,
+    PeerPool,
+)
+from p2p.utils import sxor
+
+
+class Server:
+    """Server listening for incoming connections"""
+    logger = logging.getLogger("p2p.server.Server")
+
+    def __init__(self,
+                 privkey: datatypes.PrivateKey,
+                 server_address: Address,
+                 chaindb: AsyncChainDB
+                 ) -> None:
+        self.cancel_token = CancelToken('Server')
+        self.chaindb = chaindb
+        self.privkey = privkey
+        self.server_address = server_address
+        discovery = DiscoveryProtocol(self.privkey, self.server_address, bootstrap_nodes=[])
+        self.peer_pool = PeerPool(ETHPeer, self.chaindb, 1, self.privkey, discovery)
+
+    async def run(self) -> None:
+        self.logger.info("Running server...")
+        loop = asyncio.get_event_loop()
+        factory = asyncio.start_server(
+            self.receive_handshake, host=self.server_address.ip, port=self.server_address.udp_port)
+        asyncio.ensure_future(factory)
+
+        while not self.cancel_token.triggered:
+            try:
+                loop.run_forever()
+            except OperationCancelled:
```

Comment:
This will never raise OperationCancelled. You can just `await self.cancel_token.wait()` if all you want is to yield and only resume (although you'd immediately return) once the token is triggered.